hardly/handlers/distgit.py

```diff
@@ -99,35 +107,115 @@ def run(self) -> TaskResults:
+            package_config=package_config,
+            job_config=job_config,
+            event=event,
+        )
+        # It would ideally be taken from package_config.dist_git_namespace
+        # instead of from getenv, but package_config is None because there's no config file.
+        self.src_git_namespace: str = getenv("DISTGIT_NAMESPACE").replace(
+            "/rpms", "/src"
+        )
+        # project name is expected to be the same in dist-git and src-git
+        self.src_git_name: str = event["project_name"]
+        # branch name
+        self.git_ref: str = event["git_ref"]
+
+        self.status: str = event["status"]
+        self.detailed_status: str = event["detailed_status"]
+        self.pipeline_url: str = (
+            f"{event['project_url']}/-/pipelines/{event['pipeline_id']}"
+        )
+
+        # lazy
+        self._src_git_project: Optional[GitProject] = None
+        self._src_git_mr_id: Optional[int] = None
+        self._src_git_mr: Optional[PullRequest] = None
+        self._status_reporter: Optional[StatusReporter] = None
+
+    @property
+    def src_git_mr_id(self) -> Optional[int]:
+        """
+        https://docs.gitlab.com/ee/user/project/integrations/webhooks.html#pipeline-events
+        suggests, there's a merge_request field containing relation to the dist-git MR,
+        sadly, in reality it's empty.
+        Luckily, we've stored the src-git MR in the branch name (self.git_ref)
+        from which the dist-git MR is created.
+        See how we set local_pr_branch_suffix in DistGitMRHandler.run()
+        :return: src-git MR number or None if self.git_ref doesn't contain it
+        """
+        if not self._src_git_mr_id:
+            # git_ref is expected in a form {version}-{dist_git_branch}-src-{mr_number}
+            m = fullmatch(r".+-.+-src-(\d+)", self.git_ref)
+            self._src_git_mr_id = int(m[1]) if m else None
+        return self._src_git_mr_id
+
+    @property
+    def src_git_project(self) -> GitProject:
+        if not self._src_git_project:
+            self._src_git_project = self.project.service.get_project(
+                namespace=self.src_git_namespace,
+                repo=self.src_git_name,
+            )
+        return self._src_git_project
+
+    @property
+    def src_git_mr(self) -> PullRequest:
+        if not self._src_git_mr:
+            self._src_git_mr = self.src_git_project.get_pr(self.src_git_mr_id)
+        return self._src_git_mr
+
+    @property
+    def status_reporter(self) -> StatusReporter:
+        if not self._status_reporter:
+            self._status_reporter = StatusReporter.get_instance(
+                project=self.src_git_project,
+                # The head_commit is latest commit of the MR.
+                # If there was a new commit pushed before the pipeline ended, the report
+                # might be incorrect until the new (for the new commit) pipeline finishes.
+                commit_sha=self.src_git_mr.head_commit,
```

Comment:
can we store the commit info in our database so we don't hit this race condition?

Reply:
as a fallowup work